src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,801 @@
+            }
+            sqlTx.commit();
+        }
+        catch (...)
+        {
+            mRoot->getCache().clear();
+            throw;
+        }
+
+        mRoot->hasChild(false);
+        mRoot = nullptr;
+    }
+
+    mHeader.reset();
+    mState.clear();
+    mLoadBestOfferContext.clear();
+}
+
+void
+LedgerState::mergeStateIntoParent()
+{
+    for (auto const& state : mState)
+    {
+        state.second->invalidate();
+        auto const& entry = state.second->ignoreInvalid().entry();
+
+        auto iter = mParent->mState.find(state.first);
+        if (iter == mParent->mState.end())
+        {
+            StateEntry ler(new LedgerEntryReference(entry, nullptr));
+            ler->invalidate();
+            mParent->mState[state.first] = ler;
+        }
+        else
+        {
+            auto const& previous =
+                iter->second->ignoreInvalid().previousEntry();
+            StateEntry ler(new LedgerEntryReference(entry, previous));
+            ler->invalidate();
+            iter->second = ler;
+        }
+    }
+}
+
+void
+LedgerState::mergeHeaderIntoParent()
+{
+    if (mHeader)
+    {
+        auto const& header = mHeader->ignoreInvalid().header();
+        auto const& previous = mHeader->ignoreInvalid().previousHeader();
+        StateHeader lhr(new LedgerHeaderReference(header, previous));
+        lhr->invalidate();
+        mParent->mHeader = lhr;
+    }
+}
+
+void
+LedgerState::mergeStateIntoRoot()
+{
+    for (auto const& state : mState)
+    {
+        state.second->invalidate();
+        if (state.second->ignoreInvalid().entry())
+        {
+            storeInDatabase(state.second);
+        }
+        else
+        {
+            deleteFromDatabase(state.second);
+        }
+    }
+}
+
+void
+LedgerState::mergeHeaderIntoRoot()
+{
+    if (mHeader)
+    {
+        storeHeaderInDatabase();
+        mRoot->setPreviousHeader(mHeader->ignoreInvalid().header());
+    }
+}
+
+void
+LedgerState::rollback()
+{
+    assert(mParent || mRoot);
+
+    if (mChild)
+    {
+        mChild->rollback();
+    }
+
+    if (mParent)
+    {
+        mParent->mChild = nullptr;
+        mParent = nullptr;
+    }
+    else
+    {
+        mRoot->hasChild(false);
+        mRoot = nullptr;
+    }
+
+    mHeader.reset();
+    mState.clear();
+    mLoadBestOfferContext.clear();
+}
+
+LedgerState::StateEntry
+LedgerState::create(LedgerEntry const& entry)
+{
+    assert(mParent || mRoot);
+    assert(!mChild);
+
+    auto key = LedgerEntryKey(entry);
+    auto ler = createHelper(entry, key);
+    mState[key] = ler;
+
+    if (entry.data.type() == OFFER)
+    {
+        OfferEntry const& offer = entry.data.offer();
+        invalidateLoadBestOfferContext(offer.selling, offer.buying);
+    }
+
+    return ler;
+}
+
+LedgerState::StateEntry
+LedgerState::createHelper(LedgerEntry const& entry, LedgerKey const& key)
+{
+    StateEntry ler;
+
+    auto iter = mState.find(key);
+    if (iter != mState.end())
+    {
```

Comment:
when can this happen? it is unclear to me

Reply:
Here is an example where this occurs:
```
LedgerKey key = ...;
LedgerState lsParent(...);
LedgerState lsChild(lsParent);
lsChild.load(key)->erase();
lsChild.commit();
lsParent.create(key);
```
Let me know if you need further explanation or examples on what is happening here, I'm happy to explain further.